codalab/server/bundle_manager.py

```diff
@@ -553,14 +561,49 @@ def _deduct_worker_resources(self, workers_list, running_bundles_info):
                 worker['memory_bytes'] -= bundle_resources.memory
         return workers_list
 
+    @staticmethod
+    def _create_bundle_resource_snapshot(bundle_resources):
+        """
+        Given a RunResources instance, returns a tuple of requested bundle
+        resources including CPUs, GPUs, memory and disk space.
+        """
+        cpus = bundle_resources.cpus
+        gpus = bundle_resources.gpus
+        memory = formatting.size_str(bundle_resources.memory)
+        disk = formatting.size_str(bundle_resources.disk)
+        return f'({cpus}, {gpus}, {memory}, {disk})'
+
+    @staticmethod
+    def _create_workers_resource_snapshot(workers_list):
+        """
+        Given a list of workers, returns a tuple of worker resource information
+        including CPUs, GPUs, memory and disk space for each worker.
```

Comment:
There's also the tag of the worker (corresponding to `--request-queue` on the job) that we should include.